compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -578,6 +580,183 @@ static int64_t getMaxVectorTileSize(int64_t numElem, int64_t tileSize,
+  // (which is a harder requirement).
+  for (int loopNum : llvm::reverse(llvm::seq<int>(0, numLoops))) {
+    // Skip 0 vecTileSizes, want to preserve them as 0.
+    if (vecTileSizes[loopNum] == 0) {
+      continue;
+    }
+    for (int i : llvm::seq<int>(0, numOperands)) {
+      // Check if this operand is concerned with this loop.
+      if (!indexingMaps[i].isFunctionOfDim(loopNum)) {
+        continue;
+      }
+      int64_t oldVal = vecTileSizes[loopNum];
+      int64_t maxVal = std::max(minVecTileSizes[loopNum],
+                                eachOperandMaxTileBits / tileBits[i]);
+      int64_t adjustedVal = std::min(oldVal, maxVal);
+      // If we are adjusting a tile size, make sure that the adjusted tile size
+      // is a power-of-two, as introducing non-power-of-two tile sizes can be a
+      // major performance regression, e.g. a 30% e2e regression in MobileNetV2
+      // just from a single occurrence in a DepthwiseConv (example discussed
+      // below in this comment).
+      //
+      // We round the adjustedVal to the nearest power of two /below/ not
+      // /above/ because doing that could make the entire adjustment a
+      // no-operation (e.g. from 8 to 5 back to 8) and then would trigger the
+      // tail recursion with halved eachOperandMaxTileBits, which would result
+      // in more aggressive tile size limitation. E.g. in MobileNetV2 on AVX-512
+      // we have a DepthwiseConv with
+      //   vecTileSizes = [1 1 8 16 1 3],
+      // without this adjustment, this becomes
+      //   vecTileSizes = [1 1 5 16 1 3].
+      // If we round this 5 back to 8, letting this function tail-recurse, we
+      // end up with this much narrower final result:
+      //   vecTileSizes = [1 1 2 16 1 3].
+      // It is best to round this 5 to 4, obtaining
+      //   vecTileSizes = [1 1 4 16 1 3].
+      //
+      // The above example also shows that some non-power-of-two tile sizes
+      // are not getting adjusted, and should not be rounded to nearest
+      // power-of-two: the inner-most dimension size 3 above.
+      if (adjustedVal != oldVal) {
+        // Round to nearest power of 2, rounding down.
+        adjustedVal = 1L << llvm::Log2_64(adjustedVal);
+      }
+      vecTileSizes[loopNum] = adjustedVal;
+      tileBits[i] *= adjustedVal;
+    }
+  }
+
+  // At this point we have satisfied our entry-wise requirements on
+  // `vecTileSizes`, but there is still the possibility that the sum of all
+  // operand tiles' bit sizes exceeds `allOperandsMaxTileBits`. Tail-recurse
+  // with halved `eachOperandMaxTileBits` until that requirement is satisfied.
+  // Note that if `eachOperandMaxTileBits` falls below some element type bit
+  // width, it will trigger an early-return above, so we don't need to worry
+  // about that here.
+  if (sum(tileBits) > allOperandsMaxTileBits) {
+    limitVectorTileSizes(op, vecTileSizes, eachOperandMaxTileBits / 2,
+                         allOperandsMaxTileBits);
+  }
```

Comment:
nit: I think we can use `std::accumulate`, then we don't need the sum function. E.g.,

```
int64_t totalTileBits = std::accumulate(tileBits.begin(), tileBits.end(), 0);
```

Reply:
Thanks, I couldn't remember what was the standard library utility for this. In fact, looking around, I found https://en.cppreference.com/w/cpp/algorithm/reduce which is equivalent here just a few characters less as it doesn't require passing the 0, and generally we don't care about the order of accumulation here.